### ISSUE TITLE
return of C99 compound literal addresses doesn't yield warnings

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#8678]
|  |  |
| --- | --- |
| Bugzilla Link | [8306](https://llvm.org/bz8306) |
| Version | trunk |
| OS | Linux |
| Reporter | LLVM Bugzilla Contributor |

## Extended Description 
The following code warns about returning the adress of stack memory
```cpp
struct stat *foo(void)
{
    struct stat st;

    return &st;
}
```
The following code doesn't:
```cpp
struct stat *foo(void)
{
    return &(struct stat){ 0, };
}
```
or this one either:
```cpp
void (^blk_t)(void);

blk_t foo(void) { return ^{ (void)0; } };
```
This is somehow similar to the issues in #8675 (but in worse because here the programmer is really used to the compiler reporting these mistakes)

**Comment by JOE1994** [comment]
Still reproducible with `clang 17.0.1`: https://godbolt.org/z/rdbz9svMj

**Comment by shafik** [comment]
Interesting to note gcc also does not catch this case: https://godbolt.org/z/WqsKccvz8
